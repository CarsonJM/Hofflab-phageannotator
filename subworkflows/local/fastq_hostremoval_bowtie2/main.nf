//
// Remove host reads with bowtie2
//
include { BOWTIE2_BUILD } from '../../modules/nf-core/bowtie2/build/main'
include { BOWTIE2_ALIGN } from '../../modules/nf-core/bowtie2/build/main'

workflow FASTQ_HOSTREMOVAL_BOWTIE2 {
    take:
    fastq_gz    // channel: [ [ meta.id, meta.group ], [ reads_1.fastq.gz, reads_2.fastq.gz ] ]

    main:
    ch_versions = Channel.empty()

    // identify bowtie2 index or fasta to use for host removal
    if (params.host_genome) {
        host_fasta              = params.genomes[params.host_genome].fasta ?: false
        ch_host_fasta           = Channel
            .value(
                file("${host_fasta}", checkIfExists: true)
            )
        host_bowtie2index       = params.genomes[params.host_genome].bowtie2 ?: false
        ch_host_bowtie2index    = Channel
            .value(
                file("${host_bowtie2index}/*", checkIfExists: true)
            )
    } else if ( params.host_fasta ) {
        ch_host_fasta           = Channel
            .value(
                file("${params.host_fasta}", checkIfExists: true)
            )
    }

    //
    // MODULE: Create bowtie2 database from host FASTA
    //
    if (!ch_host_bowtie2index) {
        ch_host_bowtie2index    = BOWTIE2_BUILD (ch_host_fasta).index
        ch_versions             = ch_versions.mix(BOWTIE2_BUILD.out.versions)
    }

    //
    // MODULE: Align reads to bowtie2 index
    //
    ch_bt2_reads_fastq_gz       = BOWTIE2_ALIGN (ch_short_reads_prepped,ch_host_bowtie2index).reads
    ch_versions                 = ch_versions.mix(BOWTIE2_HOST_REMOVAL_ALIGN.out.versions)

    emit:
    bt2_reads_fastq_gz          = ch_bt2_reads_fastq_gz // channel: [ [ meta.id, meta.group ], [ reads_1.fastq.gz, reads_2.fastq.gz ] ]
    versions                    = ch_versions           // [ versions.yml ]
}
