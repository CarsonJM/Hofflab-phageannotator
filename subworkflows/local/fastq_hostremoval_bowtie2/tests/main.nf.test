nextflow_workflow {

    name "Test Subworkflow: FASTQ_HOSTREMOVAL_BOWTIE2"
    script "../main.nf"
    workflow "FASTQ_HOSTREMOVAL_BOWTIE2"

    test("fastq.gz paired + fasta.gz + no index") {
        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test1', single_end:false ], // meta map
                        [
                            file("${projectDir}/assets/test-datasets/fastqs/test1_1.fastq.gz", checkifExists: true),
                            file("${projectDir}/assets/test-datasets/fastqs/test1_2.fastq.gz", checkifExists: true)
                        ]
                    ],
                    [
                        [ id:'test2', single_end:false ], // meta map
                        [
                            file("${projectDir}/assets/test-datasets/fastqs/test2_1.fastq.gz", checkifExists: true),
                            file("${projectDir}/assets/test-datasets/fastqs/test2_2.fastq.gz", checkifExists: true)
                        ]
                    ]
                )
                input[1] = [
                    [ id:'fasta' ],
                    file("${projectDir}/assets/test-datasets/fastas/lambda.fasta.gz", checkifExists: true)
                ]
                input[2] = null
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("fastq.gz paired + fasta.gz + bt2") {
        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test1', single_end:false ], // meta map
                        [
                            file("${projectDir}/assets/test-datasets/fastqs/test1_1.fastq.gz", checkifExists: true),
                            file("${projectDir}/assets/test-datasets/fastqs/test1_2.fastq.gz", checkifExists: true)
                        ]
                    ],
                    [
                        [ id:'test2', single_end:false ], // meta map
                        [
                            file("${projectDir}/assets/test-datasets/fastqs/test2_1.fastq.gz", checkifExists: true),
                            file("${projectDir}/assets/test-datasets/fastqs/test2_2.fastq.gz", checkifExists: true)
                        ]
                    ]
                )
                input[1] = [
                    [ id:'fasta' ],
                    file("${projectDir}/assets/test-datasets/bowtie2/lambda.fasta.gz", checkifExists: true)
                ]
                input[2] = [
                    [ id: 'index' ],
                    "${projectDir}/assets/test-datasets/bowtie2/"
                ]
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
