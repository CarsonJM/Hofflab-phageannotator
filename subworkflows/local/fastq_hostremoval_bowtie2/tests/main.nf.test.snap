{
    "fastq.gz paired + fasta.gz + no index": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        [
                            "test1_bowtie2.unmapped_1.fastq.gz:md5,6924a0e4bf760ad3ad6a4a5ab0020ce1",
                            "test1_bowtie2.unmapped_2.fastq.gz:md5,bc64507de9b0fafce1757cae39ff3e7a"
                        ]
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        [
                            "test2_bowtie2.unmapped_1.fastq.gz:md5,40d41a369939ab5eaefd14813594eceb",
                            "test2_bowtie2.unmapped_2.fastq.gz:md5,0323f20f5ce1d80c7f9cfaeb19b462ea"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,7527172e0c4466b67ac3676f720ad480",
                    "versions.yml:md5,a2f4bc7f555b5b1c334c97961644d42b"
                ],
                "2": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        "test1_bowtie2.bowtie2.log:md5,a947ef5d3656c605d3266370a2690275"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2_bowtie2.bowtie2.log:md5,69e4dad267a8ffd19ad21c294bc9433b"
                    ]
                ],
                "fastq_gz": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        [
                            "test1_bowtie2.unmapped_1.fastq.gz:md5,6924a0e4bf760ad3ad6a4a5ab0020ce1",
                            "test1_bowtie2.unmapped_2.fastq.gz:md5,bc64507de9b0fafce1757cae39ff3e7a"
                        ]
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        [
                            "test2_bowtie2.unmapped_1.fastq.gz:md5,40d41a369939ab5eaefd14813594eceb",
                            "test2_bowtie2.unmapped_2.fastq.gz:md5,0323f20f5ce1d80c7f9cfaeb19b462ea"
                        ]
                    ]
                ],
                "mqc": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        "test1_bowtie2.bowtie2.log:md5,a947ef5d3656c605d3266370a2690275"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2_bowtie2.bowtie2.log:md5,69e4dad267a8ffd19ad21c294bc9433b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,7527172e0c4466b67ac3676f720ad480",
                    "versions.yml:md5,a2f4bc7f555b5b1c334c97961644d42b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-20T10:45:20.478514699"
    },
    "fastq.gz paired + fasta.gz + bt2": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        [
                            "test1_bowtie2.unmapped_1.fastq.gz:md5,6924a0e4bf760ad3ad6a4a5ab0020ce1",
                            "test1_bowtie2.unmapped_2.fastq.gz:md5,bc64507de9b0fafce1757cae39ff3e7a"
                        ]
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        [
                            "test2_bowtie2.unmapped_1.fastq.gz:md5,40d41a369939ab5eaefd14813594eceb",
                            "test2_bowtie2.unmapped_2.fastq.gz:md5,0323f20f5ce1d80c7f9cfaeb19b462ea"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,a2f4bc7f555b5b1c334c97961644d42b"
                ],
                "2": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        "test1_bowtie2.bowtie2.log:md5,a947ef5d3656c605d3266370a2690275"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2_bowtie2.bowtie2.log:md5,69e4dad267a8ffd19ad21c294bc9433b"
                    ]
                ],
                "fastq_gz": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        [
                            "test1_bowtie2.unmapped_1.fastq.gz:md5,6924a0e4bf760ad3ad6a4a5ab0020ce1",
                            "test1_bowtie2.unmapped_2.fastq.gz:md5,bc64507de9b0fafce1757cae39ff3e7a"
                        ]
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        [
                            "test2_bowtie2.unmapped_1.fastq.gz:md5,40d41a369939ab5eaefd14813594eceb",
                            "test2_bowtie2.unmapped_2.fastq.gz:md5,0323f20f5ce1d80c7f9cfaeb19b462ea"
                        ]
                    ]
                ],
                "mqc": [
                    [
                        {
                            "id": "test1",
                            "single_end": false
                        },
                        "test1_bowtie2.bowtie2.log:md5,a947ef5d3656c605d3266370a2690275"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2_bowtie2.bowtie2.log:md5,69e4dad267a8ffd19ad21c294bc9433b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a2f4bc7f555b5b1c334c97961644d42b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-20T10:51:46.525642255"
    }
}